Benchmark OutputStreamWriter against OutputStream

carterkozak opened the pull request:
Initial results:
```
Benchmark                                       (charsetName)    (msg)  Mode  Cnt    Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8    ASCII  avgt    4   13.926 ±  0.908  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8  UNICODE  avgt    4  110.787 ±  2.740  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8    ASCII  avgt    4  105.864 ± 13.620  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8  UNICODE  avgt    4  262.389 ±  7.248  ns/op
```

cl4es commented:
I copied this, turned the text blocks into regular strings and compiled it for --release 8 to be able to compare back.

On my aging workstation (SandyBridge, AVX2) I get these numbers on some not very rigorous runs:

JDK 8:
```
Benchmark                                       (charsetName)    (msg)  Mode  Cnt     Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8    ASCII  avgt    4   416.184 ± 35.805  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8  UNICODE  avgt    4   886.206 ± 68.719  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8    ASCII  avgt    4   451.775 ± 29.702  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8  UNICODE  avgt    4  1107.296 ± 17.345  ns/op
```
JDK 11:
```
Benchmark                                       (charsetName)    (msg)  Mode  Cnt     Score     Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8    ASCII  avgt    4    53.722 ±  16.381  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8  UNICODE  avgt    4   542.948 ± 138.590  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8    ASCII  avgt    4   430.063 ±  60.449  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8  UNICODE  avgt    4  2024.860 ± 188.798  ns/op
```

JDK 17:
```
Benchmark                                       (charsetName)    (msg)  Mode  Cnt     Score     Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8    ASCII  avgt    4    56.065 ±   2.758  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8  UNICODE  avgt    4   583.998 ±  39.936  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8    ASCII  avgt    4   493.675 ±  63.620  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8  UNICODE  avgt    4  1121.760 ± 299.786  ns/op
```

The biggest change over time is how JDK 11 and 17 is much faster on `getBytesToOutputStream`, especially for `ASCII`. There's also a largish regression in 11 for the `outputStreamWriter` benchmark for `UNICODE` that seem to have been fixed for JDK 17.

Running with the patch I'm preparing to add ASCII fast-paths (https://github.com/openjdk/jdk/pull/5621) I get these numbers:
```
Benchmark                                       (charsetName)    (msg)  Mode  Cnt    Score     Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8    ASCII  avgt    4   59.785 ±   5.136  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream          UTF-8  UNICODE  avgt    4  561.139 ±  60.300  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8    ASCII  avgt    4  167.183 ±  36.628  ns/op
EncodedOutputBenchmarks.outputStreamWriter              UTF-8  UNICODE  avgt    4  955.499 ± 139.580  ns/op
```

This improves `outputStreamWriter` by roughly 3x for `ASCII` and maybe 1.1-1.2x for `UNICODE` (too big of an error here to say for sure), and clearly ahead of JDK 8 on all benchmarks. The remaining difference between `outputStreamWriter` and `getBytesToOutputStream` is probably unrelated to the actual character encoding, but I haven't profiled it to see what it does differently.

carterkozak commented:
My results building your branch from source on https://github.com/openjdk/jdk/pull/5621/commits/b74fdb8c0eefe044dd1510379bfa6fa3736d754b using a  Xeon W-2175 workstation.

Raw performance encoding from StringBuilder is neck-and-neck in my results, certainly close enough that I'm not worried, and from the logging library perspective it can be helpful to avoid allocation interactions which vary by workload and configuration. This is fantastic!

I realized I haven't included benchmarks encoding a StringBuilder which has been inflated to UTF16 but has been updated with only ascii characters. I suspect that will be common as we reuse StringBuilders for formatting, so eventually they'll become inflated.

run 1
```
Benchmark                                                      (charsetName)                          (message)    (msg)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A    ASCII              N/A  avgt    4   14.240 ±  1.803  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A  UNICODE              N/A  avgt    4  122.190 ± 11.964  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A    ASCII              N/A  avgt    4   50.376 ±  5.967  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A  UNICODE              N/A  avgt    4  234.628 ± 40.138  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message      N/A                3  avgt    4   25.685 ±  3.379  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  130.762 ± 10.811  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message      N/A                3  avgt    4   31.229 ± 10.570  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  131.767 ± 20.888  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message      N/A                3  avgt    4  231.086 ± 32.001  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  299.930 ± 42.848  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message      N/A                3  avgt    4   22.597 ±  0.205  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  115.271 ±  8.981  ns/op
```

run 3
```
Benchmark                                                      (charsetName)                          (message)    (msg)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A    ASCII              N/A  avgt    4   14.925 ±  1.676  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A  UNICODE              N/A  avgt    4  124.752 ± 20.187  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A    ASCII              N/A  avgt    4   49.372 ±  3.110  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A  UNICODE              N/A  avgt    4  222.911 ±  3.883  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message      N/A                3  avgt    4   19.685 ±  6.541  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  121.500 ±  5.831  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message      N/A                3  avgt    4   30.256 ±  2.917  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  126.980 ±  7.460  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message      N/A                3  avgt    4  230.210 ± 18.923  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  283.801 ±  3.595  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message      N/A                3  avgt    4   22.492 ±  1.329  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  110.280 ±  5.955  ns/op
```

vs azul zulu jdk-17 (ideally I'd build the base commit and test that)

run 2
```
Benchmark                                                      (charsetName)                          (message)    (msg)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A    ASCII              N/A  avgt    4   13.950 ±  0.128  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A  UNICODE              N/A  avgt    4  115.452 ± 14.071  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A    ASCII              N/A  avgt    4  107.887 ± 30.591  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A  UNICODE              N/A  avgt    4  309.156 ± 33.300  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message      N/A                3  avgt    4   66.456 ±  5.077  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  149.279 ± 21.112  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message      N/A                3  avgt    4   78.064 ± 10.819  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  161.745 ± 17.906  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message      N/A                3  avgt    4  275.825 ± 25.656  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  382.733 ± 54.529  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message      N/A                3  avgt    4   24.474 ±  2.660  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  122.905 ±  4.977  ns/op
```

run 4:
```
Benchmark                                                      (charsetName)                          (message)    (msg)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A    ASCII              N/A  avgt    4   14.183 ±  0.141  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                                N/A  UNICODE              N/A  avgt    4  115.273 ±  3.109  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A    ASCII              N/A  avgt    4  110.677 ± 11.558  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                                N/A  UNICODE              N/A  avgt    4  264.937 ± 28.096  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message      N/A                3  avgt    4   59.690 ± 17.378  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  130.386 ±  3.079  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message      N/A                3  avgt    4   66.683 ±  4.626  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  133.899 ±  3.502  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message      N/A                3  avgt    4  256.827 ± 24.129  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  328.158 ± 12.441  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message      N/A                3  avgt    4   22.123 ±  1.447  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊      N/A                3  avgt    4  107.383 ± 19.547  ns/op
```

carterkozak commented:
I've updated the branch to test using an inflated StringBuilder (coder=UTF16) with single-byte characters, and updated with the latest commit on your branch 8edc228f64f.
The charsetEncoder method has been consistently outperforming toString.getBytes in my testing, there may have been some background noise on my system this morning.

```
Benchmark                                                      (charsetName)               (message)    (msg)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                     N/A    ASCII              N/A  avgt    4   14.723 ±  1.181  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                     N/A  UNICODE              N/A  avgt    4  115.057 ± 13.081  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                     N/A    ASCII              N/A  avgt    4   50.895 ±  3.168  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                     N/A  UNICODE              N/A  avgt    4  204.916 ±  4.780  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8                   ASCII      N/A                3  avgt    4   20.196 ±  6.738  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8                 UNICODE      N/A                3  avgt    4  120.023 ±  8.539  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4   21.317 ±  2.223  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8                   ASCII      N/A                3  avgt    4   29.219 ±  1.543  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8                 UNICODE      N/A                3  avgt    4  131.840 ±  2.997  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4   30.984 ±  6.210  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8                   ASCII      N/A                3  avgt    4  208.991 ± 24.003  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8                 UNICODE      N/A                3  avgt    4  295.220 ± 36.173  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4  240.016 ± 16.152  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8                   ASCII      N/A                3  avgt    4   23.278 ±  1.045  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8                 UNICODE      N/A                3  avgt    4  112.128 ±  2.606  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4   26.575 ±  0.705  ns/op
```

vs jdk-17

```
Benchmark                                                      (charsetName)               (message)    (msg)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                     N/A    ASCII              N/A  avgt    4   15.249 ±  1.772  ns/op
EncodedOutputBenchmarks.getBytesToOutputStream                         UTF-8                     N/A  UNICODE              N/A  avgt    4  120.228 ± 11.898  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                     N/A    ASCII              N/A  avgt    4  106.189 ± 22.290  ns/op
EncodedOutputBenchmarks.outputStreamWriter                             UTF-8                     N/A  UNICODE              N/A  avgt    4  278.147 ± 44.254  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8                   ASCII      N/A                3  avgt    4   60.597 ±  3.338  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8                 UNICODE      N/A                3  avgt    4  130.564 ±  0.225  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4   59.744 ±  2.595  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8                   ASCII      N/A                3  avgt    4   66.975 ±  1.370  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8                 UNICODE      N/A                3  avgt    4  141.215 ±  1.940  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4   69.813 ±  1.356  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8                   ASCII      N/A                3  avgt    4  238.629 ±  4.623  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8                 UNICODE      N/A                3  avgt    4  337.577 ± 73.664  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4  263.239 ± 14.295  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8                   ASCII      N/A                3  avgt    4   23.056 ±  2.446  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8                 UNICODE      N/A                3  avgt    4  109.534 ±  1.928  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  ASCII_INFLATED_BUILDER      N/A                3  avgt    4   26.300 ±  0.552  ns/op
```

For what it's worth, I don't think we need to index heavily on the OutputStreamWriter. I selected that example as a common use-case which relies on the CharsetEncoder, and would be improved by CharsetEncoder optimizations.

cl4es commented:
That too good to be true feeling is strong right now.. 😄 

Also good to see that an inflated `StringBuilder` only hurts performance by <10%, so my intuition that this might be a significant part of it was way off. At one point I was thinking we could speculatively reset the `coder` to `LATIN1` when calling `setLength(0)`, but that's probably one of those hacks that might be worthwhile in your micro but hurt other use-cases - generally it seems reasonable to assume that a builder that has seen unicode codepoints before will see it again soon enough. With the cost of doing what you do on an inflated builder isn't prohibitively costly then perhaps we should avoid mucking too much with that.

FWIW the reason the `outputStreamWriter` variant drops off a bit seem to be that it converts the `String` to a newly allocated `char[]` in `sun.nio.cs.StreamEncoder`:

```java
    public void write(String str, int off, int len) throws IOException {
        /* Check the len before creating a char buffer */
        if (len < 0)
            throw new IndexOutOfBoundsException();
        char[] cbuf = new char[len];
        str.getChars(off, off + len, cbuf, 0);
        write(cbuf, 0, len);
    }
```
.. then that `char[]` will be wrapped by a `CharBuffer` and go through a `CharsetEncoder` to write to an internal `ByteBuffer` before flushing from that buffer to the `outputStream`. So it allocates a bit more than twice as much for ASCII, has to inflate from latin-1 to char, and copies data around one extra time.

carterkozak commented:
>  At one point I was thinking we could speculatively reset the coder to LATIN1 when calling setLength(0), but that's probably one of those hacks that might be worthwhile in your micro but hurt other use-cases - generally it seems reasonable to assume that a builder that has seen unicode codepoints before will see it again soon enough. 

I'd actually put together a commit doing just this the other day, but abandoned the idea because inflate doesn't (and I imagine can't, using simd?) run in-place. The cost of the additional array allocations and copies seems likely to be more expensive than the cost of encoding from expanded arrays.

regarding `sun.nio.cs.StreamEncoder`: I suppose we could keep around a CharBuffer based on the ByteBuffer size to avoid reallocations, that would bring us close to the charsetEncoder example in the benchmark. I wonder if we'd be better served adding a new CharsetEncoder method along the lines of:

```java
/**
 * Returns -1 if no characters were written because the destination buffer does not have enough room,
 * otherwise the number of characters encoded to destination.
 * @implNote implementations should special-case String and AbstractStringBuilder to transcode their internal byte-arrays.
 */
int encode(CharSequence text, int offset, int length, ByteBuffer destination);
```
